datasets/flwr_datasets/partitioner/semantic_partitioner.py

```diff
@@ -0,0 +1,493 @@
+    algorithm to enforce intra-client intra-label similarity and intra-client
+    inter-label similarity in two separate stages.
+
+    • Stage 1: For each label, we embed examples using a pretrained neural
+    network (extracting semantic features), and fit a Gaussian Mixture Model
+    to cluster pretrained embeddings into groups. Note that this results
+    in multiple groups per label. This stage enforces intra-client
+    intra-label consistency.
+
+    • Stage 2: To package the clusters from different labels into clients,
+    we aim to compute an optimal multi-partite matching with cost-matrix
+    defined by KL-divergence between the Gaussian clusters. To reduce complexity,
+    we heuristically solve the optimal multi-partite matching by progressively
+    solving the optimal bipartite matching at each time for randomly-chosen
+    label pairs. This stage enforces intra-client inter-label consistency.
+
+    Parameters
+    ----------
+    num_partitions : int
+        The total number of partitions that the data will be divided into.
+    partition_by : str
+        Column name of the labels (targets) based on which Dirichlet sampling works.
+    efficient_net_type: int
+        The type of pretrained EfficientNet model.
+        Options: [0, 1, 2, 3, 4, 5, 6, 7], corresponding to EfficientNet B0-B7 models.
+    batch_size: int
+        The batch size for EfficientNet extracting embeddings.
+    pca_components: int
+        The number of PCA components for dimensionality reduction.
+    gmm_max_iter: int
+        The maximum number of iterations for the GMM algorithm.
+    gmm_init_params: str
+        The initialization method for the GMM algorithm.
+        Options: ["random", "kmeans", "k-means++"]
+    use_cuda: bool
+        Whether to use CUDA for computation acceleration.
+    shuffle: bool
+        Whether to randomize the order of samples. Shuffling applied after the
+        samples assignment to partitions.
+    seed: int
+        Seed used for dataset shuffling. It has no effect if `shuffle` is False.
+
+    Examples
+    --------
+    >>> from flwr_datasets import FederatedDataset
+    >>> from flwr_datasets.partitioner import SemanticPartitioner
+    >>> partitioner = SemanticPartitioner(
+    >>>     num_partitions=10,
+    >>>     partition_by="label",
+    >>>     pca_components=128,
+    >>>     gmm_max_iter=100,
+    >>>     gmm_init_params="kmeans",
+    >>>     use_cuda=True,
+    >>>     shuffle=True,
+    >>> )
+    >>> fds = FederatedDataset(dataset="mnist", partitioners={"train": partitioner})
+    >>> partition = fds.load_partition(0)
+    >>> print(partition[0])  # Print the first example
+    {'image': <PIL.PngImagePlugin.PngImageFile image mode=L
+    size=28x28 at 0x7FCF49741B10>, 'label': 3}
+    >>> partition_sizes = partition_sizes = [
+    >>>     len(fds.load_partition(partition_id)) for partition_id in range(5)
+    >>> ]
+    >>> print(sorted(partition_sizes))
+    [3163, 5278, 5496, 6320, 9522]
```

Comment:
A more general comment: I'm profiling the code to understand better what takes time (I'll check the memory perf, too). For data bigger than 1k, it takes some time on the local machine. We should signal that progress is being made (not to confuse people that the code froze). The lack of a proper logger makes it hard, so I'll suggest a print for now.